audformat/core/utils.py

```diff
@@ -1143,6 +1143,117 @@ def set_index_dtypes(
+    :class:`pandas.Index`
+    with a single-level
+    :class:`pandas.MultiIndex`,
+    is calculated,
+    the result is a
+    :class:`pandas.Index`.
+
+    Args:
+        objs: index objects
+
+    Returns:
+        symmetric difference of index objects
+
+    Raises:
+        ValueError: if level and dtypes of objects do not match
+
+    Example:
+        >>> symmetric_difference(
+        ...     [
+        ...         pd.Index([1, 2, 3], name='idx'),
+        ...     ]
+        ... )
+        Int64Index([1, 2, 3], dtype='int64', name='idx')
+        >>> symmetric_difference(
+        ...     [
+        ...         pd.Index([0, 1], name='idx'),
+        ...         pd.Index([1, 2], dtype='Int64', name='idx'),
+        ...     ]
+        ... )
+        Index([0, 2], dtype='Int64', name='idx')
+        >>> symmetric_difference(
+        ...     [
+        ...         pd.Index([0, 1], name='idx'),
+        ...         pd.MultiIndex.from_arrays([[1, 2]], names=['idx']),
+        ...     ]
+        ... )
+        Int64Index([0, 2], dtype='int64', name='idx')
+        >>> symmetric_difference(
+        ...     [
+        ...         pd.MultiIndex.from_arrays(
+        ...             [['a', 'b', 'c'], [0, 1, 2]],
+        ...             names=['idx1', 'idx2'],
+        ...         ),
+        ...         pd.MultiIndex.from_arrays(
+        ...             [['b', 'c'], [1, 3]],
+        ...             names=['idx1', 'idx2'],
+        ...         ),
+        ...     ]
+        ... )
+        MultiIndex([('a', 0),
+                    ('c', 2),
+                    ('c', 3)],
+                   names=['idx1', 'idx2'])
+        >>> symmetric_difference(
+        ...     [
+        ...         filewise_index(['f1', 'f2', 'f3']),
+        ...         filewise_index(['f2', 'f3', 'f4']),
+        ...     ]
+        ... )
+        Index(['f1', 'f4'], dtype='string', name='file')
+        >>> symmetric_difference(
+        ...     [
+        ...         segmented_index(['f1'], [0], [1]),
+        ...         segmented_index(['f1', 'f2'], [0, 1], [1, 2]),
+        ...     ]
+        ... )
+        MultiIndex([('f2', '0 days 00:00:01', '0 days 00:00:02')],
+                   names=['file', 'start', 'end'])
+        >>> symmetric_difference(
+        ...     [
+        ...         filewise_index(['f1', 'f2']),
+        ...         segmented_index(['f1', 'f2'], [0, 0], [pd.NaT, 1]),
+        ...     ]
+        ... )
+        MultiIndex([('f2', '0 days',               NaT),
+                    ('f2', '0 days', '0 days 00:00:01')],
+                   names=['file', 'start', 'end'])
+
+    """
+    if not objs:
+        return pd.Index([])
+
+    if len(objs) == 1:
+        return objs[0]
+
+    objs = _maybe_convert_filewise_index(objs)
+    objs = _maybe_convert_single_level_multi_index(objs)
+
+    index = union(objs)
+    count = np.zeros(len(index))
+    for obj in objs:
+        count += index.isin(obj)
+    index = index[count == 1]
```

Comment:
I was again wondering if it would be possible to implement this without relying on `union()` to safe time and memory for very large indices.

The current implementation requires 4.4 s on the same data I used in https://github.com/audeering/audformat/pull/234#discussion_r931901348.

If you change the code to relying on actual set operations
```python
    difference = set()
    for obj in objs:
        difference = difference ^ set(obj)
    difference = tuple(difference)
    if isinstance(obj, pd.MultiIndex):
        index = pd.MultiIndex.from_tuples(
            difference,
            names=obj.names,
        )
        index = set_index_dtypes(index, dict(obj.dtypes))
    else:
        index = pd.Index(difference, name=obj.name, dtype=obj.dtype)
```
it doesn't read that nice any longer, but it runs in 0.35 s.

Another side effect is that some tests are failing as you now get the results back in a different order.

Reply:
Maybe the the `if`-statement could also be moved to a helper function.

Reply:
I did another check, we can speed up `intersect()` by using `set()` as well to 0.1 s, but our implementation for `union()` is already very fast and will be slower if we switch to `set()` there as well.

Reply:
Ok, I suggest we stay with the current solution and merge here so that I can continue working on https://github.com/audeering/audformat/pull/225 and do the speed up using `set` here and in `intersect()` in another PR.

Reply:
Good idea, I created https://github.com/audeering/audformat/issues/245 and will take a look at it after merging this.